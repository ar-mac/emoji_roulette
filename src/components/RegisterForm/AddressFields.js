import React, { Component } from 'react';
import PropTypes from 'proptypes';
import { get } from 'lodash';

export class AddressFields extends Component {
  render() {
    const { address, index, removeAddress, handleCityChange, handleChange } = this.props;
    return (
      <div
        className="panel panel-default" key={index}
        role="group"
        aria-labelledby={`panel-title-${index}`}
      >
        <div className="panel-heading">
          <h3
            className="panel-title"
            id={`panel-title-${index}`}
          >
            address #{index + 1}
          </h3>
          <button onClick={removeAddress}>Remove</button>
        </div>
        <div className="panel-body">
          <div className={cn('form-group', {
            'has-error': get(errors, `addresses[${index}].city`)
          })}>
            <label htmlFor={`addresses[${index}].city`}>city</label>
            <select
              className="form-control"
              name={`addresses[${index}].city`}
              id={`addresses[${index}].city`}
              value={address.city}
              onChange={handleCityChange}
            >
              <option value=""></option>
              <optgroup label="Polska">
                <option value="Sosnowiec">Sosnowiec</option>
                <option value="Bielsko-Biała">Bielsko-Biała</option>
                <option value="Cieszyn">Cieszyn</option>
              </optgroup>
              <optgroup label="Europa">
                <option value="Czarnobyl">Czarnobyl</option>
                <option value="Berlin">Berlin</option>
                <option value="Praga">Praga</option>
                <option value="Zylin">Zylin</option>
              </optgroup>
            </select>
            {get(errors, `addresses[${index}].city`) &&
            <span className="help-block">{errors.addresses[index].city}</span>
            }
          </div>
          <div className={cn('form-group', {
            'has-error': get(errors, `addresses[${index}].zipCode`)
          })}>
            <label htmlFor={`addresses[${index}].zipCode`}>zip-code</label>
            <input
              type="text"
              className="form-control"
              name={`addresses[${index}].zipCode`}
              id={`addresses[${index}].zipCode`}
              aria-describedby="zipcode_format"
              placeholder="Provide your zip-code"
              pattern="\d{2}-\d{3}"
              value={address.zipCode}
              onChange={handleChange}
            />
            {get(errors, `addresses[${index}].zipCode`) &&
            <span className="help-block">{errors.addresses[index].zipCode}</span>
            }
            <div className="visuallyhidden" id="zipcode_format">00-000</div>
          </div>
        </div>
      </div>
    )
  }
}

AddressFields.propTypes = {
  address: PropTypes.object,
  errors: PropTypes.object,
  index: PropTypes.number,
  removeAddress: PropTypes.func,
  handleCityChange: PropTypes.func,
  handleChange: PropTypes.func,
};
