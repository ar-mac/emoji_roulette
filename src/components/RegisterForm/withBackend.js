import React, { Component } from 'react';
import PropTypes from 'proptypes';
import { isEmpty } from 'lodash';
import { connect } from 'react-redux';

import { login, logout } from '../../store/registration/actionCreators';

const usernames = ['user', 'user1', 'Bob', 'Me'];

export const withBackend = (WrappedComponent) => {
  class WithBackend extends Component {
    handleSubmitByBackend = (formData) => {
      return new Promise((resolve, reject) => {
        setTimeout(() => {
          const errors = this.validateByBackend(formData);
          console.log('Dispatched logout', errors);                     
          if (isEmpty(errors)) {
            resolve();
            this.props.login();
          } else {
            reject(errors);
            this.props.logout();
          }
        }, 200);
      })
    };

    validateByBackend = (formData) => {
      const errors = {};
      if (!formData.username) {
        errors.username = 'Is required';
      }
      if (formData.username && usernames.includes(formData.username)) {
        errors.username = 'Is already taken';
      }

      return errors
    };

    render() {
      return <WrappedComponent handleSubmit={this.handleSubmitByBackend} />
    }
  }

  WithBackend.displayName = `withBackend(${WrappedComponent.displayName || 'Component'})`;

  WithBackend.propTypes = {
    login: PropTypes.func.isRequired,
    logout: PropTypes.func.isRequired,
  };

  const mapDispatchToProps = { login, logout };
  return connect(null, mapDispatchToProps)(WithBackend);
};
