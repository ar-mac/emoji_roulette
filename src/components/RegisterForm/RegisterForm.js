import React, { Component } from 'react';
import PropTypes from 'proptypes';
import { get, set, find, cloneDeep } from 'lodash';
import cn from 'classnames';
import yup from 'yup';

import { withBackend } from "./withBackend";
import { AddressFields } from "./AddressFields";

const cities = [
  { city: 'Sosnowiec', zipCode: '00-000' },
  { city: 'Bielsko-Biała', zipCode: '43-300' },
  { city: 'Cieszyn', zipCode: '43-430' },
  { city: 'Czarnobyl', zipCode: '66-666' },
  { city: 'Berlin', zipCode: '30-821' },
  { city: 'Praga', zipCode: '04-022' },
  { city: 'Zylin', zipCode: '82-742' },
];

export class RegisterForm extends Component {
  state = {
    data: {
      age: '',
      username: '',
      email: '',
      addresses: [
        { city: '', zipCode: '' },
      ]
    },
    errors: {},
    isSending: false,
  };

  schema = yup.object().shape({
    age: yup.string().required(),
    username: yup.string().required().min(3).max(50),
    email: yup.string().email(),
    addresses: yup.array().of(yup.object().shape({
      city: yup.string(),
      zipCode: yup.string(),
    }))
  });

  submit = (event) => {
    event.preventDefault();

    const { data } = this.state;
    this.schema.validate(data)
      .then(isValid => {
        if (isValid) {
          this.setState({ isSending: true });
          this.props.handleSubmit(data).catch(errors => this.setState({ errors }));
        }
      })
      .catch(errors => this.setState({ errors }))
  };

  handleChange = ({ target }) => {
    this.setState((currentState) => {
      const newData = cloneDeep(currentState.data);
      set(newData, target.name, target.value);
      return { data: newData };
    });
  };

  removeAddress = (index) => () => {
    const { data } = this.state;

    const newData = {
      ...data,
      addresses: [
        ...data.addresses.slice(0, index),
        ...data.addresses.slice(index + 1),
      ]
    };

    this.setState({ data: newData });
  };

  addAddress = () => {
    const { data } = this.state;

    const newData = {
      ...data,
      addresses: [
        ...data.addresses,
        { city: '', zipCode: '' },
      ]
    };

    this.setState({ data: newData });
  };

  handleCityChange = (index) => (event) => {
    this.handleChange(event);

    const city = find(cities, { city: event.target.value });
    if (city) {
      this.handleChange({ target: { value: city.zipCode, name: `addresses[${index}].zipCode` } })
    }
  };

  render() {
    const { data, errors } = this.state;
    return (
      <div className="register-form row col-md-6 col-md-offset-3">
        <form onSubmit={this.submit}>
          <div className={cn('form-group', {
            'has-error': get(errors, 'age')
          })}>
            <label htmlFor="age">age</label>
            <input
              type="number"
              className="form-control"
              name="age"
              id="age"
              placeholder="Provide your age"
              value={data.age}
              onChange={this.handleChange}
            />
            {get(errors, 'age') &&
            <span className="help-block">{errors.age}</span>
            }
          </div>
          <div className={cn('form-group', {
            'has-error': get(errors, 'username')
          })}>
            <label htmlFor="username">username</label>
            <input
              type="text"
              className="form-control"
              name="username"
              id="username"
              placeholder="Provide your username"
              value={data.username}
              onChange={this.handleChange}
            />
            {get(errors, 'username') &&
            <span className="help-block">{errors.username}</span>
            }
          </div>
          <div className={cn('form-group', {
            'has-error': get(errors, 'email')
          })}>
            <label htmlFor="email">email</label>
            <input
              type="email"
              className="form-control"
              name="email"
              id="email"
              placeholder="Provide your email"
              value={data.email}
              onChange={this.handleChange}
            />
            {get(errors, 'email') &&
            <span className="help-block">{errors.email}</span>
            }
          </div>
          {data.addresses.map((address, index) => (
            <AddressFields
              key={index}
              address={address}
              index={index}
              errors={errors}
              removeAddress={this.removeAddress(index)}
              handleCityChange={this.handleCityChange(index)}
              handleChange={this.handleChange}
            />
          ))}
          <div>
            <button onClick={this.addAddress}>Add new address</button>
          </div>
          <br />
          ᕙ(◔◡◔)ᕗ
          <div>
            <button>Submit form</button>
          </div>
        </form>
      </div>
    )
  }
}

RegisterForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
};

export default withBackend(RegisterForm);
