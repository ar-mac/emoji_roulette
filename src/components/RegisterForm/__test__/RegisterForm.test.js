import React from 'react';

import { RegisterForm } from '../RegisterForm';

describe('RegisterForm', () => {
  const defaultProps = {
    handleSubmit: jest.fn(),
  };

  const setup = buildSetup(RegisterForm, defaultProps);

  test('renders with default props', () => {
    expect(setup().wrapper).toMatchSnapshot();
  });
});
