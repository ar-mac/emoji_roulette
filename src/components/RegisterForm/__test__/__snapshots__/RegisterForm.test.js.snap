// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm renders with default props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RegisterForm
    handleSubmit={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
>
        <div
                className="form-group"
        >
                <label
                        htmlFor="age"
                >
                        age
                </label>
                <input
                        className="form-control"
                        id="age"
                        name="age"
                        onChange={[Function]}
                        placeholder="Provide your age"
                        type="number"
                        value=""
                />
        </div>
        <div
                className="form-group"
        >
                <label
                        htmlFor="username"
                >
                        username
                </label>
                <input
                        className="form-control"
                        id="username"
                        name="username"
                        onChange={[Function]}
                        placeholder="Provide your username"
                        type="text"
                        value=""
                />
        </div>
        <div
                className="form-group"
        >
                <label
                        htmlFor="email"
                >
                        email
                </label>
                <input
                        className="form-control"
                        id="email"
                        name="email"
                        onChange={[Function]}
                        placeholder="Provide your email"
                        type="email"
                        value=""
                />
        </div>
        <AddressFields
                address={
                        Object {
                                "city": "",
                                "zipCode": "",
                              }
                }
                errors={Object {}}
                handleChange={[Function]}
                handleCityChange={[Function]}
                index={0}
                removeAddress={[Function]}
        />
        <div>
                <button
                        onClick={[Function]}
                >
                        Add new address
                </button>
        </div>
        <br />
        ᕙ(◔◡◔)ᕗ
        <div>
                <button>
                        Submit form
                </button>
        </div>
</form>,
      "className": "register-form row col-md-6 col-md-offset-3",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="form-group"
>
            <label
                        htmlFor="age"
            >
                        age
            </label>
            <input
                        className="form-control"
                        id="age"
                        name="age"
                        onChange={[Function]}
                        placeholder="Provide your age"
                        type="number"
                        value=""
            />
</div>,
          <div
            className="form-group"
>
            <label
                        htmlFor="username"
            >
                        username
            </label>
            <input
                        className="form-control"
                        id="username"
                        name="username"
                        onChange={[Function]}
                        placeholder="Provide your username"
                        type="text"
                        value=""
            />
</div>,
          <div
            className="form-group"
>
            <label
                        htmlFor="email"
            >
                        email
            </label>
            <input
                        className="form-control"
                        id="email"
                        name="email"
                        onChange={[Function]}
                        placeholder="Provide your email"
                        type="email"
                        value=""
            />
</div>,
          Array [
            <AddressFields
              address={
                            Object {
                                          "city": "",
                                          "zipCode": "",
                                        }
              }
              errors={Object {}}
              handleChange={[Function]}
              handleCityChange={[Function]}
              index={0}
              removeAddress={[Function]}
/>,
          ],
          <div>
            <button
                        onClick={[Function]}
            >
                        Add new address
            </button>
</div>,
          <br />,
          "ᕙ(◔◡◔)ᕗ",
          <div>
            <button>
                        Submit form
            </button>
</div>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label
                htmlFor="age"
>
                age
</label>,
              <input
                className="form-control"
                id="age"
                name="age"
                onChange={[Function]}
                placeholder="Provide your age"
                type="number"
                value=""
/>,
              undefined,
            ],
            "className": "form-group",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "age",
                "htmlFor": "age",
              },
              "ref": null,
              "rendered": "age",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "className": "form-control",
                "id": "age",
                "name": "age",
                "onChange": [Function],
                "placeholder": "Provide your age",
                "type": "number",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            undefined,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label
                htmlFor="username"
>
                username
</label>,
              <input
                className="form-control"
                id="username"
                name="username"
                onChange={[Function]}
                placeholder="Provide your username"
                type="text"
                value=""
/>,
              undefined,
            ],
            "className": "form-group",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "username",
                "htmlFor": "username",
              },
              "ref": null,
              "rendered": "username",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "className": "form-control",
                "id": "username",
                "name": "username",
                "onChange": [Function],
                "placeholder": "Provide your username",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            undefined,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label
                htmlFor="email"
>
                email
</label>,
              <input
                className="form-control"
                id="email"
                name="email"
                onChange={[Function]}
                placeholder="Provide your email"
                type="email"
                value=""
/>,
              undefined,
            ],
            "className": "form-group",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "email",
                "htmlFor": "email",
              },
              "ref": null,
              "rendered": "email",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "className": "form-control",
                "id": "email",
                "name": "email",
                "onChange": [Function],
                "placeholder": "Provide your email",
                "type": "email",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            undefined,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "address": Object {
              "city": "",
              "zipCode": "",
            },
            "errors": Object {},
            "handleChange": [Function],
            "handleCityChange": [Function],
            "index": 0,
            "removeAddress": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": <button
              onClick={[Function]}
>
              Add new address
</button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Add new address",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add new address",
            "type": "button",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        "ᕙ(◔◡◔)ᕗ",
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": <button>
              Submit form
</button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Submit form",
            },
            "ref": null,
            "rendered": "Submit form",
            "type": "button",
          },
          "type": "div",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
>
          <div
                    className="form-group"
          >
                    <label
                              htmlFor="age"
                    >
                              age
                    </label>
                    <input
                              className="form-control"
                              id="age"
                              name="age"
                              onChange={[Function]}
                              placeholder="Provide your age"
                              type="number"
                              value=""
                    />
          </div>
          <div
                    className="form-group"
          >
                    <label
                              htmlFor="username"
                    >
                              username
                    </label>
                    <input
                              className="form-control"
                              id="username"
                              name="username"
                              onChange={[Function]}
                              placeholder="Provide your username"
                              type="text"
                              value=""
                    />
          </div>
          <div
                    className="form-group"
          >
                    <label
                              htmlFor="email"
                    >
                              email
                    </label>
                    <input
                              className="form-control"
                              id="email"
                              name="email"
                              onChange={[Function]}
                              placeholder="Provide your email"
                              type="email"
                              value=""
                    />
          </div>
          <AddressFields
                    address={
                              Object {
                                        "city": "",
                                        "zipCode": "",
                                      }
                    }
                    errors={Object {}}
                    handleChange={[Function]}
                    handleCityChange={[Function]}
                    index={0}
                    removeAddress={[Function]}
          />
          <div>
                    <button
                              onClick={[Function]}
                    >
                              Add new address
                    </button>
          </div>
          <br />
          ᕙ(◔◡◔)ᕗ
          <div>
                    <button>
                              Submit form
                    </button>
          </div>
</form>,
        "className": "register-form row col-md-6 col-md-offset-3",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="form-group"
>
              <label
                            htmlFor="age"
              >
                            age
              </label>
              <input
                            className="form-control"
                            id="age"
                            name="age"
                            onChange={[Function]}
                            placeholder="Provide your age"
                            type="number"
                            value=""
              />
</div>,
            <div
              className="form-group"
>
              <label
                            htmlFor="username"
              >
                            username
              </label>
              <input
                            className="form-control"
                            id="username"
                            name="username"
                            onChange={[Function]}
                            placeholder="Provide your username"
                            type="text"
                            value=""
              />
</div>,
            <div
              className="form-group"
>
              <label
                            htmlFor="email"
              >
                            email
              </label>
              <input
                            className="form-control"
                            id="email"
                            name="email"
                            onChange={[Function]}
                            placeholder="Provide your email"
                            type="email"
                            value=""
              />
</div>,
            Array [
              <AddressFields
                address={
                                Object {
                                                "city": "",
                                                "zipCode": "",
                                              }
                }
                errors={Object {}}
                handleChange={[Function]}
                handleCityChange={[Function]}
                index={0}
                removeAddress={[Function]}
/>,
            ],
            <div>
              <button
                            onClick={[Function]}
              >
                            Add new address
              </button>
</div>,
            <br />,
            "ᕙ(◔◡◔)ᕗ",
            <div>
              <button>
                            Submit form
              </button>
</div>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <label
                  htmlFor="age"
>
                  age
</label>,
                <input
                  className="form-control"
                  id="age"
                  name="age"
                  onChange={[Function]}
                  placeholder="Provide your age"
                  type="number"
                  value=""
/>,
                undefined,
              ],
              "className": "form-group",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "children": "age",
                  "htmlFor": "age",
                },
                "ref": null,
                "rendered": "age",
                "type": "label",
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "className": "form-control",
                  "id": "age",
                  "name": "age",
                  "onChange": [Function],
                  "placeholder": "Provide your age",
                  "type": "number",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              undefined,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <label
                  htmlFor="username"
>
                  username
</label>,
                <input
                  className="form-control"
                  id="username"
                  name="username"
                  onChange={[Function]}
                  placeholder="Provide your username"
                  type="text"
                  value=""
/>,
                undefined,
              ],
              "className": "form-group",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "children": "username",
                  "htmlFor": "username",
                },
                "ref": null,
                "rendered": "username",
                "type": "label",
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "className": "form-control",
                  "id": "username",
                  "name": "username",
                  "onChange": [Function],
                  "placeholder": "Provide your username",
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              undefined,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <label
                  htmlFor="email"
>
                  email
</label>,
                <input
                  className="form-control"
                  id="email"
                  name="email"
                  onChange={[Function]}
                  placeholder="Provide your email"
                  type="email"
                  value=""
/>,
                undefined,
              ],
              "className": "form-group",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "children": "email",
                  "htmlFor": "email",
                },
                "ref": null,
                "rendered": "email",
                "type": "label",
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "className": "form-control",
                  "id": "email",
                  "name": "email",
                  "onChange": [Function],
                  "placeholder": "Provide your email",
                  "type": "email",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              undefined,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "class",
            "props": Object {
              "address": Object {
                "city": "",
                "zipCode": "",
              },
              "errors": Object {},
              "handleChange": [Function],
              "handleCityChange": [Function],
              "index": 0,
              "removeAddress": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": <button
                onClick={[Function]}
>
                Add new address
</button>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Add new address",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add new address",
              "type": "button",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          "ᕙ(◔◡◔)ᕗ",
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": <button>
                Submit form
</button>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Submit form",
              },
              "ref": null,
              "rendered": "Submit form",
              "type": "button",
            },
            "type": "div",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
